File: javascript/word-count/word-count.js
module.exports = function words(source) {
  return lowercaseWords(source).reduce(function(counts, word) {
    counts[word] = (counts[word] || 0) + 1;
    return counts;
  }, Object.create(null));
};

function lowercaseWords(str) {
  return str.toLowerCase().match(/\w+/g);
}

File: clojure/word-count/word_count.clj
(ns word-count
  (require [clojure.string :as str]))

(def non-word-characters #"[^\p{L}\p{N}]+")

(defn- split-into-lowercase-words [string]
  (str/split (str/lower-case string) non-word-characters))

(defn- count-word [counts word]
  (assoc counts word (inc (get counts word 0))))

(defn word-count [string]
  (reduce count-word {} (split-into-lowercase-words string)))
